toolchain/check/interface.cpp

```diff
@@ -40,55 +40,112 @@ auto BuildAssociatedEntity(Context& context, SemIR::InterfaceId interface_id,
+// Given a `Self` type and a witness that it implements an interface, along with
+// that interface's `Self` binding, form a facet that can be used as the
+// argument for that `Self` binding.
+static auto GetSelfFacet(Context& context,
+                         SemIR::SpecificId interface_specific_id,
+                         SemIR::InstId self_binding_id,
+                         SemIR::TypeId self_type_id,
+                         SemIR::InstId self_witness_id) -> SemIR::InstId {
+  // Check that we were given a self binding. The binding might be a
+  // `BindSymbolicName` or an `ImportRef` naming one.
+  {
+    auto self_binding_const_inst_id =
+        context.constant_values().GetConstantInstId(self_binding_id);
+    auto bind_name_inst = context.insts().GetAs<SemIR::BindSymbolicName>(
+        self_binding_const_inst_id);
+    CARBON_CHECK(
+        context.entity_names().Get(bind_name_inst.entity_name_id).name_id ==
+            SemIR::NameId::SelfType,
+        "Expected a Self binding, found {0}", bind_name_inst);
   }
 
-  // Add the `Self` argument. First find the `Self` binding.
-  auto self_binding =
-      context.insts().GetAs<SemIR::BindSymbolicName>(bindings[arg_ids.size()]);
-  CARBON_CHECK(
-      context.entity_names().Get(self_binding.entity_name_id).name_id ==
-          SemIR::NameId::SelfType,
-      "Expected a Self binding, found {0}", self_binding);
+  auto self_binding = context.insts().Get(self_binding_id);
+  auto self_facet_type_id = SemIR::GetTypeInSpecific(
+      context.sem_ir(), interface_specific_id, self_binding.type_id());
   // Create a facet value to be the value of `Self` in the interface.
   // TODO: Pass this in instead of creating it here. The caller sometimes
   // already has a facet value.
   auto type_inst_id = context.types().GetInstId(self_type_id);
-  auto facet_value_const_id =
+  auto self_value_const_id =
       TryEvalInst(context, SemIR::InstId::None,
-                  SemIR::FacetValue{.type_id = self_binding.type_id,
+                  SemIR::FacetValue{.type_id = self_facet_type_id,
                                     .type_inst_id = type_inst_id,
-                                    .witness_inst_id = witness_inst_id});
-  arg_ids.push_back(context.constant_values().GetInstId(facet_value_const_id));
+                                    .witness_inst_id = self_witness_id});
+  return context.constant_values().GetInstId(self_value_const_id);
+}
+
+static auto GetSelfFacet(Context& context,
+                         SemIR::SpecificId interface_specific_id,
+                         SemIR::GenericId generic_id,
+                         SemIR::TypeId self_type_id,
+                         SemIR::InstId self_witness_id) -> SemIR::InstId {
+  return GetSelfFacet(
+      context, interface_specific_id,
+      GetSelfBinding(context, interface_specific_id, generic_id), self_type_id,
+      self_witness_id);
+}
+
+static auto GetGenericArgsWithSelfType(Context& context,
```

Comment:
```suggestion
// Returns a new argument list with `Self` appended.
static auto GetGenericArgsWithSelfType(Context& context,
```

This is essentially just an `append`, right?